File: ios/Cookpit/Cookpit/PhotoDataController.swift
//
//  PhotoDataController.swift
//  Cookpit
//
//  Created by Kittinun Vantasin on 7/8/16.
//  Copyright © 2016 Cookpit. All rights reserved.
//

import Foundation
import RxSwift

class PhotoDetailDataController {

  private let controller: CPPhotoDetailController
  
  fileprivate let _viewData = Variable<CPPhotoDetailViewData?>(nil)
  
  lazy var viewData: Observable<CPPhotoDetailViewData> = {
    self._viewData.asObservable()
        .filter { $0 != nil }
        .map { $0! as CPPhotoDetailViewData }
        .observeOn(MainScheduler.instance)
  }()
  
  init(id: String) {
    controller = CPPhotoDetailController.create(id)!
    controller.subscribe(self)
  }
  
  func unsubscribe() {
    controller.unsubscribe()
  }
  
  func request() {
    controller.requestDetail()
  }

  deinit {
    controller.unsubscribe()
  }

}

extension PhotoDetailDataController : CPPhotoDetailControllerObserver {

  func onBeginUpdate() {
  }

  func onUpdate(_ viewData: CPPhotoDetailViewData) {
    _viewData.value = viewData
  }

  func onEndUpdate() {
  }

}

class PhotoCommentDataController {

  private let controller: CPPhotoCommentController
  
  fileprivate let _viewData = Variable<CPPhotoCommentViewData?>(nil)
  
  lazy var viewData: Observable<CPPhotoCommentViewData> = {
    self._viewData.asObservable()
        .filter { $0 != nil }
        .map { $0! as CPPhotoCommentViewData }
        .observeOn(MainScheduler.instance)
  }()
  
  init(id: String) {
    controller = CPPhotoCommentController.create(id)!
    controller.subscribe(self)
  }
  
  func unsubscribe() {
    controller.unsubscribe()
  }
  
  func request() {
    controller.requestComments()
  }

  deinit {
    controller.unsubscribe()
  }
  
}

extension PhotoCommentDataController : CPPhotoCommentControllerObserver {

  func onBeginUpdate() {

  }

  func onUpdate(_ viewData: CPPhotoCommentViewData) {
    _viewData.value = viewData
  }

  func onEndUpdate() {
        
  }

}

File: android/Cookpit/app/src/main/kotlin/com/github/kittinunf/cookpit/photo/PhotoDataController.kt
package com.github.kittinunf.cookpit.photo

import com.github.kittinunf.cookpit.*
import com.github.kittinunf.reactiveandroid.MutableProperty

class PhotoDetailDataController(id: String) : PhotoDetailControllerObserver() {

    private val controller: PhotoDetailController

    private val _viewData = MutableProperty<PhotoDetailViewData>()

    val viewData by lazy { _viewData.observable }

    private val _loadings = MutableProperty(false)

    val loadings by lazy { _loadings.observable }

    init {
        controller = PhotoDetailController.create(id)
        controller.subscribe(this)
    }

    fun unsubscribe() {
        controller.unsubscribe()
    }

    fun request() {
        controller.requestDetail()
    }

    //region PhotoDetailControllerObserver
    override fun onBeginUpdate() {
        _loadings.value = true
    }

    override fun onUpdate(viewData: PhotoDetailViewData?) {
        _viewData.value = viewData
    }

    override fun onEndUpdate() {
        _loadings.value = false
    }
    //endregion

}

class PhotoCommentDataController(id: String): PhotoCommentControllerObserver() {

    private val controller: PhotoCommentController

    private val _viewData = MutableProperty<PhotoCommentViewData>()

    val viewData by lazy { _viewData.observable }

    private val _loadings = MutableProperty(false)

    val loadings by lazy { _loadings.observable }

    init {
        controller = PhotoCommentController.create(id)
        controller.subscribe(this)
    }

    fun unsubscribe() {
        controller.unsubscribe()
    }

    fun request() {
        controller.requestComments()
    }

    //region PhotoCommentControllerObserver
    override fun onBeginUpdate() {
        _loadings.value = true
    }

    override fun onUpdate(viewData: PhotoCommentViewData?) {
        _viewData.value = viewData
    }

    override fun onEndUpdate() {
        _loadings.value = false
    }
    //endregion

}